sdks/python/apache_beam/ml/transforms/handlers.py

```diff
@@ -0,0 +1,431 @@
+
+    if is_primitive_container:
+      dtype = typing.get_args(typ)[0]  # type: ignore[attr-defined]
+      if len(typing.get_args(typ)) > 1 or typing.get_origin(dtype) == Union:  # type: ignore[attr-defined]
+        raise RuntimeError(
+            f"Union type is not supported for column: {col_name}. "
+            f"Please pass a PCollection with valid schema for column "
+            f"{col_name} by passing a single type "
+            "in container. For example, List[int].")
+    elif issubclass(typ, np.generic) or typ in _default_type_to_tensor_type_map:
+      dtype = typ
+    else:
+      raise TypeError(
+          f"Unable to identify type: {typ} specified on column: {col_name}. "
+          f"Please provide a valid type from the following: "
+          f"{_default_type_to_tensor_type_map.keys()}")
+    return tf.io.VarLenFeature(_default_type_to_tensor_type_map[dtype])
+
+  def get_raw_data_metadata(self, input_types: Dict[str, type]):
+    """
+    Return metadata to be used with tft_beam.AnalyzeAndTransformDataset
+    Args:
+      input_types: A dictionary of column names and types.
+    """
+    raise NotImplementedError
+
+  def write_transform_artifacts(self, transform_fn, location):
+    """
+    Write transform artifacts to the given location.
+    Args:
+      transform_fn: A transform_fn object.
+      location: A location to write the artifacts.
+    Returns:
+      A PCollection of WriteTransformFn writing a TF transform graph.
+    """
+    return (
+        transform_fn
+        | 'Write Transform Artifacts' >>
+        transform_fn_io.WriteTransformFn(location))
+
+  def _get_artifact_location(self, pipeline: beam.Pipeline):
+    """
+    Return the artifact location. If the pipeline options has staging location
+    set, then we will use that as the artifact location. Otherwise, we will
+    create a temporary directory and use that as the artifact location.
+    Args:
+      pipeline: A beam pipeline object.
+    Returns:
+      A location to write the artifacts.
+    """
+    # let us get the staging location from the pipeline options
+    # and initialize it as the artifact location.
+    staging_location = pipeline.options.view_as(
+        GoogleCloudOptions).staging_location
+    if not staging_location:
+      return tempfile.mkdtemp()
+    else:
+      return staging_location
+
+  def process_data_fn(
+      self, inputs: Dict[str, common_types.ConsistentTensorType]
+  ) -> Dict[str, common_types.ConsistentTensorType]:
+    """
+    A preprocessing_fn which should be implemented by subclasses
+    of TFTProcessHandlers. In this method, tft data transforms
+    such as scale_0_to_1 functions are called.
+    Args:
+      inputs: A dictionary of column names and associated data.
+    """
+    raise NotImplementedError
+
+  def _fail_on_non_gloabl_window(self, pcoll):
+    window_fn = pcoll.windowing.windowfn
+    if not isinstance(window_fn, GlobalWindows):
```

Comment:
Do we support non-trivial triggering? If not (e.g. we want to do a global combine) we should use `pcoll.windowing.is_default()`.